### ISSUE TITLE
Defining a Skylark py_proto_library via proto_library

### ISSUE THREAD

**Issue by htuch** [bazelbuild/bazel#3935]
Given https://github.com/bazelbuild/bazel/issues/2626, I was wondering whether it would be possible to add a pure Skylark `py_proto_library` built on `proto_library`.

I've played around with rules and had a look at how rules_go does this with toolchains (https://github.com/bazelbuild/rules_go/pull/803/files). These basically involve you maintaining an additional dependency tree (which `proto_library` is supposed to solve) and having your custom `py_proto_library` applied across this tree. The most promising approach to avoiding this seems to be to use aspects.

It's pretty clear how aspects could be used to invoke `protoc` to generate the `_pb2.py` files. What's less clear is how to use these outputs, which are unknown ahead of time and come from the `proto_library` transitive dep provider, in conjunction with the native `py_library` rule (I _think_ this is the sandwich problem?).

Any guidance here from the Bazel proto or Python teams?

**Comment by dslomov** [comment]
The idea here is that:
 * aspects will collect all the transitive `_pb2.py` files and pass it to `py_proto_library` rule
 * `py_proto_library` rule will make those available to `py_library` rule in the `py` provider.
This is not super-wel documented.

**Comment by htuch** [comment]
@dslomov can Bazel team do a worked example of using aspects + providers to implement a `py_proto_library` as a blog post? When I think back to the most useful docs I've seen on Bazel, I think these extended examples are great, e.g. I learned a lot about `repository_rule` from following the C++ autoconfigured toolchain example linked to from https://docs.bazel.build/versions/master/skylark/repository_rules.html.

**Comment by shariat** [comment]
any update on this would be really appreciated

**Comment by junghoahnsc** [comment]
Which rule should I use for now?
Should I use https://github.com/google/protobuf/blob/master/protobuf.bzl#L316?

**Comment by gnossen** [comment]
@dslomov 
Ping. It seems this feature is in limbo between the gRPC, Protobuf, and Bazel teams. Any updates on the Bazel end? A conversation revolving around it is unfolding at https://github.com/grpc/grpc/pull/19183.

**Comment by gnossen** [comment]
@dslomov @brandjon Friendly ping.

**Comment by brandjon** [comment]
+@hlopko since I think there's plans about prioritizing proto support in general.

**Comment by hlopko** [comment]
Yes, I'm trying to get us a person to work on proto rules. No concrete results yet, only active effort :)

**Comment by sseveran** [comment]
Any updates on this?

**Comment by hlopko** [comment]
Yes, we didn't get the person :/

**Comment by nicknezis** [comment]
Why not leverage the work being done here? It seems they are handling Python and a few more languages. https://github.com/stackb/rules_proto

**Comment by pcj** [comment]
Also worth noting that grpc folks did implement `py_proto_library` and `py_grpc_library` since earlier comments.  I've used them and they work on a routeguide example but have not used them extensively.

**Comment by awbraunstein** [comment]
What's the Bazel team's philosophy on rules like the `py_proto_library` rule being requested here. Currently there are implementations from the protobuf team, grpc team, as well as other users such as http://github.com/stackb/rules_proto mentioned above. Is the end goal for the bazel team to own these rules? Or should they be owned by some other entity?

As a user of these rules it becomes difficult to understand which one is the best/right one to use since they all tend to function in the same way (except some use the `proto_libaray` and some don't.)

**Comment by dayfine** [comment]
As far as I've tried, none of these `py_proto_library` works to my satisfaction (so are some other rules that I've tried, like `go_proto_library`). My expectation for these rules come from google3, and it's frustrating the bazel versions don't quite work the same way.

What I would expect with a `lang_proto_library` and which is not currently how these rules work:
```
proto_library(
    name = "foo_proto",
    srcs = ["foo.proto"],
    deps = [
        "bar_proto",
        "@com_google_protobuf//:timestamp_proto",
    ],
)

py_proto_library(
    name = "foo_py_pb2",
    deps = [":foo_proto"],
)

proto_library(
    name = "bar_proto",
    srcs = ["bar.proto"],
)

py_proto_library(
    name = "bar_py_pb2",
    deps = [":bar_proto"],
)

# Using foo_py_pb2 without fuss
py_library(
    deps = [":foo_py_pb2"],
)
```

Basically:
1. `lang_proto_library` only needs `foo_proto` in dependencies, not other `bar_py_pb2`. The Go rule and some flavors of the python rule do not work like this.
2. When using the proto `foo` in python, I only need `foo_py_pb2` as the dependency of the users, i.e. the target would include all its transitive dependencies. The `grpc` version of `py_proto_library` fails with this.

**Comment by scal444** [comment]
Has there been any action here? Agree with the last 2 comments, each version has its own quirks and does not align with internal Google use, which is quite frustrating when trying to maintain open-source libraries that Google also uses.

**Comment by weichunstevelee** [comment]
Hi I'm curious about the current state. Reading the current implementation of py_proto_library in protocobuffer repo, it says

```    """Deprecated alias for use before Bazel 5.3.
    Args:
      *args: the name of the py_proto_library.
      **kwargs: other keyword arguments that are passed to py_library.
    Deprecated:
      This is provided for backwards compatibility only.  Bazel 5.3 will
      introduce support for py_proto_library, which should be used instead.```

https://github.com/protocolbuffers/protobuf/blob/main/protobuf.bzl#L636

Did it happen?

**Comment by cnsgsz** [comment]
[rules_proto v2](https://github.com/stackb/rules_proto) seems to behave like @dayfine described? Is it the recommended one to use?
([rules_proto](https://github.com/bazelbuild/rules_proto) seems abandoned.)

**Comment by comius** [comment]
Starlark `py_proto_library` is available in https://github.com/bazelbuild/rules_python/blob/main/python/proto.bzl#L21
